pa-rserve/src/main/java/org/ow2/parserve/util/rsession/Rsession.java

```diff
@@ -0,0 +1,287 @@
+    public static final boolean UNIX_OPTIMIZE = true;
+    public final static String HEAD_SET = "[set] ";
+    public final static String HEAD_EVAL = "[eval] ";
+    public final static String HEAD_SERVEREVAL = "[server-eval] ";
+    public final static String HEAD_EXCEPTION = "[exception] ";
+    public final static String HEAD_ERROR = "[error] ";
+    public final static String ROUTPUT_END = "<rsession-end>";
+    public final static String IO_HEAD = "[IO] ";
+
+    private static final Logger logger = Logger.getLogger(Rsession.class);
+    public static RServeConnectionFactory rserveConnectionFactory;
+    public boolean TRY_MODE_DEFAULT = false;
+    public boolean TRY_MODE = false;
+    public RConnection connection;
+    public boolean connected = false;
+    public String name;
+    public RServeConf conf;
+
+    public String status = STATUS_NOT_SET;
+    boolean tryLocalRServe;
+
+    private String outputFile;
+
+    /**
+     * create rsession using System as a logger
+     */
+    protected Rsession(String name, RServeConf conf) {
+        this.name = name;
+        this.conf = conf;
+        RServeConnectionFactory.initializeOnce(conf);
+        begin();
+    }
+
+
+    /**
+     * Build a new Rsession, connects to an existing server or fork a new one.
+     *
+     * @param conf config of RServe
+     *             port, login, password, rProperties to pass to R (eg http_proxy)
+     */
+    public static Rsession newInstanceTry(String name, RServeConf conf) {
+        return new Rsession(name, conf);
+    }
+
+    /**
+     * @return status of Rsession
+     */
+    public String getStatus() {
+        return status;
+    }
+
+    void begin() {
+        status = STATUS_NOT_CONNECTED;
+
+        status = STATUS_CONNECTING;
+
+        connection = rserveConnectionFactory.connect();
+        connected = (connection != null);
+
+        if (!connected) {
+            status = STATUS_ERROR;
+            String message = "[" + name + "]" + "Rserve " + conf + " is not accessible.";
+            logger.error(message);
+        } else if (connection.getServerVersion() < MinRserveVersion) {
+            status = STATUS_ERROR;
+            String message = "[" + name + "]" + "Rserve " + conf + " version is too old.";
+            logger.error(message);
+        } else {
+            status = STATUS_READY;
+            return;
+        }
+
+        if (!connected) {//failed !
+
+            String message2 = "[" + name + "]" + "Failed to connect to local Rserve. Unable to initialize Rsession.";
+            logger.error(message2);
+            throw new IllegalArgumentException(message2);
+        } else {
+            logger.info("[" + name + "]" + "Connected to Local Rserve. (Version " + connection.getServerVersion() + ")");
+        }
+    }
+
+    /**
+     * correctly (depending on execution platform) shutdown Rsession.
+     */
+    public void end() {
+        if (connection == null) {
```

Comment:
This method is called only inside the Rsession constructor. Accordingly, as connection is an instance variable, I don't see how two threads could compete here.